pkg/compactor/shuffle_sharding_grouper.go

```diff
@@ -142,103 +162,134 @@ func (g *ShuffleShardingGrouper) Groups(blocks map[ulid.ULID]*metadata.Meta) (re
+			continue
+		} else if isVisited {
+			level.Info(g.logger).Log("msg", "skipping group because at least one block in group is visited", "group_hash", groupHash)
+			continue
+		}
+
+		remainingCompactions++
+		groupKey := createGroupKey(groupHash, group)
+
+		level.Info(g.logger).Log("msg", "found compactable group for user", "group_hash", groupHash, "group", group.String())
+		markBlocksVisited(g.ctx, g.bkt, g.logger, group.blocks, g.ringLifecyclerID, g.blockVisitMarkerWriteFailed)
+
+		// All the blocks within the same group have the same downsample
+		// resolution and external labels.
+		resolution := group.blocks[0].Thanos.Downsample.Resolution
+		externalLabels := labels.FromMap(group.blocks[0].Thanos.Labels)
+
+		thanosGroup, err := compact.NewGroup(
+			log.With(g.logger, "groupKey", groupKey, "rangeStart", group.rangeStartTime().String(), "rangeEnd", group.rangeEndTime().String(), "externalLabels", externalLabels, "downsampleResolution", resolution),
+			g.bkt,
+			groupKey,
+			externalLabels,
+			resolution,
+			false, // No malformed index.
+			true,  // Enable vertical compaction.
+			g.compactions.WithLabelValues(groupKey),
+			g.compactionRunsStarted.WithLabelValues(groupKey),
+			g.compactionRunsCompleted.WithLabelValues(groupKey),
+			g.compactionFailures.WithLabelValues(groupKey),
+			g.verticalCompactions.WithLabelValues(groupKey),
+			g.garbageCollectedBlocks,
+			g.blocksMarkedForDeletion,
+			g.blocksMarkedForNoCompact,
+			g.hashFunc,
+			g.blockFilesConcurrency,
+			g.blocksFetchConcurrency,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "create compaction group")
+		}
+
+		for _, m := range group.blocks {
+			if err := thanosGroup.AppendMeta(m); err != nil {
+				return nil, errors.Wrap(err, "add block to compaction group")
+			}
+		}
 
-	if len(rs.Instances) != 1 {
-		return false, fmt.Errorf("unexpected number of compactors in the shard (expected 1, got %d)", len(rs.Instances))
+		outGroups = append(outGroups, thanosGroup)
+		if len(outGroups) == g.compactionConcurrency {
+			break mainLoop
+		}
 	}
 
-	return rs.Instances[0].Addr == g.ringLifecyclerAddr, nil
+	level.Info(g.logger).Log("msg", fmt.Sprintf("total groups for compaction: %d", len(outGroups)))
+
+	return outGroups, nil
+}
+
+func (g *ShuffleShardingGrouper) isGroupVisited(blocks []*metadata.Meta) (bool, error) {
+	for _, block := range blocks {
+		blockID := block.ULID.String()
+		blockVisitMarker, err := ReadBlockVisitMarker(g.ctx, g.bkt, blockID, g.blockVisitMarkerReadFailed)
+		if err != nil {
+			if errors.Is(err, ErrorBlockVisitMarkerNotFound) {
+				level.Debug(g.logger).Log("msg", "no visit marker file for block", "blockID", blockID)
+				continue
+			}
+			level.Error(g.logger).Log("msg", "unable to read block visit marker file", "blockID", blockID, "err", err)
+			return true, err
+		}
+		if blockVisitMarker.isVisited(g.blockVisitMarkerTimeout) {
```

Comment:
Should ignore if the block was visited by the same compactor that is trying to compact now?